File: backend/services/mpesaService.js
const crypto = require('crypto');

const MPESA_ENV = process.env.MPESA_ENV || 'sandbox';
const MPESA_CONSUMER_KEY = process.env.MPESA_CONSUMER_KEY;
const MPESA_CONSUMER_SECRET = process.env.MPESA_CONSUMER_SECRET;
const MPESA_PASSKEY = process.env.MPESA_PASSKEY;
const MPESA_SHORTCODE = process.env.MPESA_SHORTCODE;
const MPESA_CALLBACK_URL = process.env.MPESA_CALLBACK_URL;
const MPESA_C2B_SHORTCODE = process.env.MPESA_C2B_SHORTCODE || MPESA_SHORTCODE;
const MPESA_C2B_RESPONSE_TYPE =
  process.env.MPESA_C2B_RESPONSE_TYPE || 'Completed';
const MPESA_C2B_CONFIRMATION_URL =
  process.env.MPESA_C2B_CONFIRMATION_URL || MPESA_CALLBACK_URL;
const MPESA_C2B_VALIDATION_URL = process.env.MPESA_C2B_VALIDATION_URL;
const MPESA_INITIATOR_NAME = process.env.MPESA_INITIATOR_NAME;
const MPESA_SECURITY_CREDENTIAL = process.env.MPESA_SECURITY_CREDENTIAL;
const MPESA_B2B_QUEUE_URL = process.env.MPESA_B2B_QUEUE_URL || MPESA_CALLBACK_URL;
const MPESA_B2B_RESULT_URL = process.env.MPESA_B2B_RESULT_URL || MPESA_CALLBACK_URL;

const BASE_URL =
  MPESA_ENV === 'production'
    ? 'https://api.safaricom.co.ke'
    : 'https://sandbox.safaricom.co.ke';

function requireEnv(name, value) {
  if (!value) {
    throw new Error(`Missing required environment variable: ${name}`);
  }
}

function formatPhone(phone) {
  const digits = phone.replace(/[^\d]/g, '');
  if (digits.startsWith('254')) return digits;
  if (digits.startsWith('0')) return `254${digits.slice(1)}`;
  if (digits.startsWith('7') && digits.length === 9) return `254${digits}`;
  if (digits.startsWith('+254')) return digits.slice(1);
  return digits;
}

function timestamp() {
  const now = new Date();
  const pad = (n) => n.toString().padStart(2, '0');
  return (
    now.getFullYear().toString() +
    pad(now.getMonth() + 1) +
    pad(now.getDate()) +
    pad(now.getHours()) +
    pad(now.getMinutes()) +
    pad(now.getSeconds())
  );
}

function stkPassword(ts) {
  // Trim any whitespace that might sneak in from .env
  const shortcode = (MPESA_SHORTCODE || '').trim();
  const passkey = (MPESA_PASSKEY || '').trim();
  const data = `${shortcode}${passkey}${ts}`;
  
  // DEBUG: Log password generation
  console.log('🔐 Password Debug:', {
    shortcodeLen: shortcode.length,
    passkeyLen: passkey.length,
    tsLen: ts.length,
    totalLen: data.length,
    rawData: data.substring(0, 20) + '...',
  });
  
  return Buffer.from(data).toString('base64');
}

async function getAccessToken() {
  requireEnv('MPESA_CONSUMER_KEY', MPESA_CONSUMER_KEY);
  requireEnv('MPESA_CONSUMER_SECRET', MPESA_CONSUMER_SECRET);

  // Trim whitespace from env vars
  const consumerKey = (MPESA_CONSUMER_KEY || '').trim();
  const consumerSecret = (MPESA_CONSUMER_SECRET || '').trim();

  const auth = Buffer.from(
    `${consumerKey}:${consumerSecret}`
  ).toString('base64');

  const res = await fetch(
    `${BASE_URL}/oauth/v1/generate?grant_type=client_credentials`,
    {
      method: 'GET',
      headers: {
        Authorization: `Basic ${auth}`,
        Accept: 'application/json',
      },
    }
  );

  if (!res.ok) {
    const body = await res.text();
    throw new Error(
      `Failed to get access token (${res.status}): ${body || res.statusText}`
    );
  }

  const data = await res.json();
  
  // DEBUG: Log the FULL raw OAuth response - NO TRUNCATION
  console.log('🎫 RAW OAUTH RESPONSE:', JSON.stringify(data, null, 2));
  console.log('🎫 FULL ACCESS TOKEN:', data.access_token);
  console.log('🎫 TOKEN LENGTH:', data.access_token?.length);
  
  if (!data.access_token) {
    throw new Error('No access_token in OAuth response: ' + JSON.stringify(data));
  }
  
  // Verify token length - should be > 100 chars
  if (data.access_token.length < 100) {
    console.warn('⚠️ WARNING: Token seems too short! Expected > 100 chars, got:', data.access_token.length);
  }
  
  return data.access_token;
}

async function initiateStkPush({ phoneNumber, amountKES, accountReference, description }) {
  requireEnv('MPESA_PASSKEY', MPESA_PASSKEY);
  requireEnv('MPESA_SHORTCODE', MPESA_SHORTCODE);
  requireEnv('MPESA_CALLBACK_URL', MPESA_CALLBACK_URL);

  const formattedPhone = formatPhone(phoneNumber);
  const accessToken = await getAccessToken();
  const ts = timestamp();
  const password = stkPassword(ts);

  // DEBUG: Log credentials being used (remove in production)
  console.log('🔑 STK Push Debug:', {
    consumerKey: MPESA_CONSUMER_KEY ? MPESA_CONSUMER_KEY.substring(0, 10) + '...' : 'NOT SET',
    shortcode: MPESA_SHORTCODE,
    passkey: MPESA_PASSKEY ? MPESA_PASSKEY.substring(0, 10) + '...' : 'NOT SET',
    timestamp: ts,
    timestampLength: ts.length,
    passwordLength: password.length,
    tokenPrefix: accessToken ? accessToken.substring(0, 10) + '...' : 'NOT SET',
    phone: formattedPhone,
  });

  // Trim env vars to avoid whitespace issues
  const shortcode = (MPESA_SHORTCODE || '').trim();
  const callbackUrl = (MPESA_CALLBACK_URL || '').trim();

  const body = {
    BusinessShortCode: shortcode,
    Password: password,
    Timestamp: ts,
    TransactionType: 'CustomerPayBillOnline',
    Amount: Math.round(amountKES).toString(),
    PartyA: formattedPhone,
    PartyB: shortcode,
    PhoneNumber: formattedPhone,
    CallBackURL: callbackUrl,
    AccountReference: accountReference,
    TransactionDesc: description || 'AKOFA Purchase',
  };

  // Daraja 3.0 STK push endpoint (processrequest)
  const url = `${BASE_URL}/mpesa/stkpush/v1/processrequest`;
  
  // DEBUG: Log the actual Authorization header being sent
  const authHeader = `Bearer ${accessToken}`;
  console.log('🔒 FINAL TOKEN CHECK:', {
    fullTokenLength: accessToken.length,
    tokenValid: accessToken.length > 100 ? '✅ VALID' : '❌ TOO SHORT',
    fullToken: accessToken,  // Log full token for debugging
  });
  
  const res = await fetch(url, {
    method: 'POST',
    headers: {
      Authorization: authHeader,
      'Content-Type': 'application/json',
      Accept: 'application/json',
    },
    body: JSON.stringify(body),
  });

  const raw = await res.text();
  let data = {};
  try {
    data = JSON.parse(raw);
  } catch (e) {
    // non-JSON response
  }

  if (!res.ok || data.ResponseCode !== '0') {
    console.error('STK push failed', {
      url,
      status: res.status,
      statusText: res.statusText,
      raw,
      data,
    });
    throw new Error(
      data.errorMessage ||
        data.errorCode ||
        data.ResponseDescription ||
        raw ||
        `STK push failed (${res.status})`
    );
  }

  return {
    checkoutRequestId: data.CheckoutRequestID,
    responseCode: data.ResponseCode,
    customerMessage: data.CustomerMessage,
    merchantRequestId: data.MerchantRequestID,
  };
}

async function queryStkStatus(checkoutRequestId) {
  requireEnv('MPESA_PASSKEY', MPESA_PASSKEY);
  requireEnv('MPESA_SHORTCODE', MPESA_SHORTCODE);

  const accessToken = await getAccessToken();
  const ts = timestamp();
  const password = stkPassword(ts);

  const body = {
    BusinessShortCode: MPESA_SHORTCODE,
    Password: password,
    Timestamp: ts,
    CheckoutRequestID: checkoutRequestId,
  };

  const res = await fetch(`${BASE_URL}/mpesa/stkpushquery/v1/query`, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${accessToken}`,
      'Content-Type': 'application/json',
      Accept: 'application/json',
    },
    body: JSON.stringify(body),
  });

  const data = await res.json().catch(() => ({}));

  if (!res.ok) {
    throw new Error(
      data.errorMessage ||
        data.errorCode ||
        data.ResultDesc ||
        JSON.stringify(data) ||
        `Query failed (${res.status})`
    );
  }

  return {
    resultCode: data.ResultCode,
    resultDesc: data.ResultDesc,
    response: data,
  };
}

/**
 * Initiate a B2B payment (BusinessBuyGoods)
 * Moves funds from your business shortcode (PartyA) to another shortcode (PartyB).
 * SecurityCredential must be the encrypted initiator password from Safaricom.
 */
async function initiateB2BPayment({
  amount,
  partyB,
  accountReference,
  remarks,
  requester,
  commandId = 'BusinessBuyGoods',
  senderIdentifierType = '4',
  receiverIdentifierType = '4',
}) {
  requireEnv('MPESA_INITIATOR_NAME', MPESA_INITIATOR_NAME);
  requireEnv('MPESA_SECURITY_CREDENTIAL', MPESA_SECURITY_CREDENTIAL);
  requireEnv('MPESA_SHORTCODE', MPESA_SHORTCODE);

  const accessToken = await getAccessToken();

  const body = {
    Initiator: MPESA_INITIATOR_NAME,
    SecurityCredential: MPESA_SECURITY_CREDENTIAL,
    CommandID: commandId,
    SenderIdentifierType: senderIdentifierType,
    RecieverIdentifierType: receiverIdentifierType,
    Amount: Math.round(Number(amount)).toString(),
    PartyA: MPESA_SHORTCODE,
    PartyB: partyB,
    AccountReference: accountReference,
    Remarks: remarks || 'B2B Payment',
    QueueTimeOutURL: MPESA_B2B_QUEUE_URL,
    ResultURL: MPESA_B2B_RESULT_URL,
    ...(requester ? { Requester: requester } : {}),
  };

  const url = `${BASE_URL}/mpesa/b2b/v1/paymentrequest`;
  const res = await fetch(url, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${accessToken}`,
      'Content-Type': 'application/json',
      Accept: 'application/json',
    },
    body: JSON.stringify(body),
  });

  const raw = await res.text();
  let data = {};
  try {
    data = JSON.parse(raw);
  } catch (e) {
    // non-JSON response
  }

  if (!res.ok || data.ResponseCode !== '0') {
    throw new Error(
      data.errorMessage ||
        data.errorCode ||
        data.ResponseDescription ||
        raw ||
        `B2B payment failed (${res.status})`
    );
  }

  return {
    originatorConversationId: data.OriginatorConversationID,
    conversationId: data.ConversationID,
    responseCode: data.ResponseCode,
    responseDescription: data.ResponseDescription,
  };
}

/**
 * Register C2B confirmation/validation URLs
 */
async function registerC2BUrls() {
  requireEnv('MPESA_C2B_SHORTCODE', MPESA_C2B_SHORTCODE);
  requireEnv('MPESA_C2B_RESPONSE_TYPE', MPESA_C2B_RESPONSE_TYPE);
  requireEnv('MPESA_C2B_CONFIRMATION_URL', MPESA_C2B_CONFIRMATION_URL);

  const accessToken = await getAccessToken();

  const body = {
    ShortCode: MPESA_C2B_SHORTCODE,
    ResponseType: MPESA_C2B_RESPONSE_TYPE, // "Completed" or "Cancelled"
    ConfirmationURL: MPESA_C2B_CONFIRMATION_URL,
    ...(MPESA_C2B_VALIDATION_URL ? { ValidationURL: MPESA_C2B_VALIDATION_URL } : {}),
  };

  const url = `${BASE_URL}/mpesa/c2b/v2/registerurl`;
  const res = await fetch(url, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${accessToken}`,
      'Content-Type': 'application/json',
      Accept: 'application/json',
    },
    body: JSON.stringify(body),
  });

  const raw = await res.text();
  let data = {};
  try {
    data = JSON.parse(raw);
  } catch (_) {}

  if (!res.ok || data.ResponseCode !== '0') {
    throw new Error(
      data.errorMessage ||
        data.errorCode ||
        data.ResponseDescription ||
        raw ||
        `C2B register URL failed (${res.status})`
    );
  }

  return {
    originatorConversationId: data.OriginatorCoversationID || data.OriginatorConversationID,
    responseCode: data.ResponseCode,
    responseDescription: data.ResponseDescription,
  };
}

/**
 * Simulate a C2B transaction (sandbox only)
 */
async function simulateC2B({
  amount,
  msisdn,
  billRefNumber,
  commandId = 'CustomerPayBillOnline',
}) {
  requireEnv('MPESA_C2B_SHORTCODE', MPESA_C2B_SHORTCODE);

  const accessToken = await getAccessToken();

  const body = {
    ShortCode: Number(MPESA_C2B_SHORTCODE),
    CommandID: commandId, // CustomerPayBillOnline or CustomerBuyGoodsOnline
    Amount: Math.round(Number(amount)),
    Msisdn: msisdn,
    BillRefNumber: billRefNumber || '',
  };

  const url = `${BASE_URL}/mpesa/c2b/v2/simulate`;
  const res = await fetch(url, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${accessToken}`,
      'Content-Type': 'application/json',
      Accept: 'application/json',
    },
    body: JSON.stringify(body),
  });

  const raw = await res.text();
  let data = {};
  try {
    data = JSON.parse(raw);
  } catch (_) {}

  if (!res.ok || data.ResponseCode !== '0') {
    throw new Error(
      data.errorMessage ||
        data.errorCode ||
        data.ResponseDescription ||
        raw ||
        `C2B simulate failed (${res.status})`
    );
  }

  return {
    originatorConversationId: data.OriginatorCoversationID || data.OriginatorConversationID,
    responseCode: data.ResponseCode,
    responseDescription: data.ResponseDescription,
  };
}

module.exports = {
  initiateStkPush,
  queryStkStatus,
  initiateB2BPayment,
  registerC2BUrls,
  simulateC2B,
};

File: lib/services/mpesa_service.dart
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import 'package:firebase_auth/firebase_auth.dart';
import 'package:cloud_firestore/cloud_firestore.dart';

class MpesaService {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;
  final FirebaseAuth _auth = FirebaseAuth.instance;
  
  // M-Pesa API endpoints
  static const String _baseUrl = 'https://sandbox.safaricom.co.ke';
  static const String _authUrl = '$_baseUrl/oauth/v1/generate?grant_type=client_credentials';
  // Daraja 3.0 uses the /process endpoint (processrequest was deprecated)
  static const String _stkPushUrl = '$_baseUrl/mpesa/stkpush/v1/process';
  static const String _queryUrl = '$_baseUrl/mpesa/stkpushquery/v1/query';
  
  // M-Pesa credentials - in production, these should be stored securely
  static const String _consumerKey = 'YOUR_CONSUMER_KEY';
  static const String _consumerSecret = 'YOUR_CONSUMER_SECRET';
  static const String _passKey = 'YOUR_PASS_KEY';
  static const String _shortCode = 'YOUR_SHORT_CODE';
  static const String _callbackUrl = 'YOUR_CALLBACK_URL';
  
  // Get OAuth token
  Future<String> _getAccessToken() async {
    final auth = 'Basic ${base64Encode(utf8.encode('$_consumerKey:$_consumerSecret'))}';
    
    try {
      final response = await http.get(
        Uri.parse(_authUrl),
        headers: {
          'Authorization': auth,
          'Accept': 'application/json',
        },
      );
      
      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return data['access_token'];
      } else {
        throw Exception('Failed to get access token: ${response.body}');
      }
    } catch (e) {
      throw Exception('Error getting access token: $e');
    }
  }
  
  // Generate timestamp in the format required by M-Pesa
  String _generateTimestamp() {
    final now = DateTime.now();
    final year = now.year.toString();
    final month = now.month.toString().padLeft(2, '0');
    final day = now.day.toString().padLeft(2, '0');
    final hour = now.hour.toString().padLeft(2, '0');
    final minute = now.minute.toString().padLeft(2, '0');
    final second = now.second.toString().padLeft(2, '0');
    
    return '$year$month$day$hour$minute$second';
  }
  
  // Generate password for STK Push
  String _generatePassword(String timestamp) {
    final dataToEncode = '$_shortCode$_passKey$timestamp';
    return base64Encode(utf8.encode(dataToEncode));
  }
  
  // Initiate STK Push
  Future<Map<String, dynamic>> initiateSTKPush(String phoneNumber, double amount, String accountReference) async {
    try {
      if (kDebugMode) {
      }
      
      // Format phone number (remove leading 0 or +254)
      if (phoneNumber.startsWith('0')) {
        phoneNumber = '254${phoneNumber.substring(1)}';
      } else if (phoneNumber.startsWith('+254')) {
        phoneNumber = phoneNumber.substring(1);
      }
      
      // Get access token
      final accessToken = await _getAccessToken();
      
      // Generate timestamp
      final timestamp = _generateTimestamp();
      
      // Generate password
      final password = _generatePassword(timestamp);
      
      // Prepare request body
      final body = {
        'BusinessShortCode': _shortCode,
        'Password': password,
        'Timestamp': timestamp,
        'TransactionType': 'CustomerPayBillOnline',
        'Amount': amount.round().toString(),
        'PartyA': phoneNumber,
        'PartyB': _shortCode,
        'PhoneNumber': phoneNumber,
        'CallBackURL': _callbackUrl,
        'AccountReference': accountReference,
        'TransactionDesc': 'Purchase of AKOFA tokens',
      };
      
      // Make API request
      final response = await http.post(
        Uri.parse(_stkPushUrl),
        headers: {
          'Authorization': 'Bearer $accessToken',
          'Content-Type': 'application/json',
          'Accept': 'application/json',
        },
        body: json.encode(body),
      );
      
      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        
        if (kDebugMode) {
        }
        
        // Store transaction details in Firestore
        await _recordMpesaTransaction(
          phoneNumber,
          amount,
          data['CheckoutRequestID'],
          accountReference,
          'pending'
        );
        
        return {
          'success': true,
          'checkoutRequestId': data['CheckoutRequestID'],
          'responseCode': data['ResponseCode'],
          'customerMessage': data['CustomerMessage'],
        };
      } else {
        if (kDebugMode) {
        }
        
        return {
          'success': false,
          'error': 'Failed to initiate payment: ${response.body}',
        };
      }
    } catch (e) {
      if (kDebugMode) {
      }
      
      return {
        'success': false,
        'error': 'Error initiating payment: $e',
      };
    }
  }
  
  // Query STK Push status
  Future<Map<String, dynamic>> querySTKStatus(String checkoutRequestId) async {
    try {
      // Get access token
      final accessToken = await _getAccessToken();
      
      // Generate timestamp
      final timestamp = _generateTimestamp();
      
      // Generate password
      final password = _generatePassword(timestamp);
      
      // Prepare request body
      final body = {
        'BusinessShortCode': _shortCode,
        'Password': password,
        'Timestamp': timestamp,
        'CheckoutRequestID': checkoutRequestId,
      };
      
      // Make API request
      final response = await http.post(
        Uri.parse(_queryUrl),
        headers: {
          'Authorization': 'Bearer $accessToken',
          'Content-Type': 'application/json',
          'Accept': 'application/json',
        },
        body: json.encode(body),
      );
      
      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        
        if (kDebugMode) {
        }
        
        // Update transaction status in Firestore
        if (data['ResultCode'] == '0') {
          await _updateMpesaTransactionStatus(checkoutRequestId, 'completed');
          
          // Credit the user's account with AKOFA tokens
          await _creditUserAccount(checkoutRequestId);
        } else {
          await _updateMpesaTransactionStatus(checkoutRequestId, 'failed');
        }
        
        return {
          'success': true,
          'resultCode': data['ResultCode'],
          'resultDesc': data['ResultDesc'],
        };
      } else {
        if (kDebugMode) {
        }
        
        return {
          'success': false,
          'error': 'Failed to query payment status: ${response.body}',
        };
      }
    } catch (e) {
      if (kDebugMode) {
      }
      
      return {
        'success': false,
        'error': 'Error querying payment status: $e',
      };
    }
  }
  
  // Record M-Pesa transaction in Firestore
  Future<DocumentReference> _recordMpesaTransaction(
    String phoneNumber,
    double amount,
    String checkoutRequestId,
    String accountReference,
    String status
  ) async {
    final String uid = _auth.currentUser!.uid;
    
    return await _firestore.collection('mpesa_transactions').add({
      'userId': uid,
      'phoneNumber': phoneNumber,
      'amount': amount,
      'checkoutRequestId': checkoutRequestId,
      'accountReference': accountReference,
      'status': status,
      'timestamp': FieldValue.serverTimestamp(),
    });
  }
  
  // Update M-Pesa transaction status in Firestore
  Future<void> _updateMpesaTransactionStatus(String checkoutRequestId, String status) async {
    final querySnapshot = await _firestore.collection('mpesa_transactions')
      .where('checkoutRequestId', isEqualTo: checkoutRequestId)
      .get();
    
    if (querySnapshot.docs.isNotEmpty) {
      await querySnapshot.docs.first.reference.update({
        'status': status,
        'updatedAt': FieldValue.serverTimestamp(),
      });
    }
  }
  
  // Credit user account with AKOFA tokens after successful payment
  Future<void> _creditUserAccount(String checkoutRequestId) async {
    try {
      // Get the transaction details
      final querySnapshot = await _firestore.collection('mpesa_transactions')
        .where('checkoutRequestId', isEqualTo: checkoutRequestId)
        .limit(1)
        .get();
      
      if (querySnapshot.docs.isEmpty) {
        if (kDebugMode) {
        }
        return;
      }
      
      final transactionDoc = querySnapshot.docs.first;
      final transactionData = transactionDoc.data();
      
      // Calculate AKOFA tokens to credit (1 AKOFA = 5.52 KES)
      final double amount = transactionData['amount'] ?? 0.0;
      final double akofaAmount = amount / 5.52;
      
      // Get user ID
      final String userId = transactionData['userId'] ?? _auth.currentUser!.uid;
      
      // Update the transaction with the AKOFA amount
      await transactionDoc.reference.update({
        'akofaAmount': akofaAmount,
        'status': 'credited',
        'creditedAt': FieldValue.serverTimestamp(),
      });
      
      // TODO: Implement actual token transfer using Stellar SDK
      // This would involve:
      // 1. Getting the issuer account credentials
      // 2. Creating a payment transaction to the user's Stellar account
      // 3. Submitting the transaction to the Stellar network
      
      if (kDebugMode) {
      }
    } catch (e) {
      if (kDebugMode) {
      }
    }
  }
  
  // Get M-Pesa transaction history for current user
  Future<List<Map<String, dynamic>>> getMpesaTransactionHistory() async {
    try {
      final String uid = _auth.currentUser!.uid;
      
      final querySnapshot = await _firestore.collection('mpesa_transactions')
        .where('userId', isEqualTo: uid)
        .orderBy('timestamp', descending: true)
        .get();
      
      return querySnapshot.docs.map((doc) {
        final data = doc.data();
        data['id'] = doc.id;
        return data;
      }).toList();
    } catch (e) {
      if (kDebugMode) {
      }
      return [];
    }
  }
}